overlord/snapstate/handlers.go

```diff
@@ -1942,6 +1942,128 @@ func (m *SnapManager) doPreferAliases(t *state.Task, _ *tomb.Tomb) error {
 	return nil
 }
 
+// nearlyReady returns whether all lanes (other than me) are Ready.
+func nearlyReady(me string, change *state.Change) bool {
+	for _, task := range change.Tasks() {
+		if me == task.ID() {
+			// ignore self
+			continue
+		}
+		if !task.Status().Ready() {
+			return false
+		}
+	}
+	return true
+}
+
+// laneSnaps returns the instance names from the first SnapSetup in each
+// non-zero Done lane in the refresh batch leading up the given task
+func laneSnaps(reTask *state.Task) []string {
+	tid := reTask.ID()
+	laneSnaps := map[int]string{}
+	// change.Tasks() preserves the order tasks were added, otherwise it all falls apart
+	for _, task := range reTask.Change().Tasks() {
+		if task.ID() == tid {
+			// we've reached ourselves; we don't care about anything beyond this
+			break
+		}
+		if task.Kind() == "check-rerefresh" {
+			// we've reached a previous check-rerefresh (but not ourselves).
+			// Only snaps in tasks after this point are of interest.
+			laneSnaps = map[int]string{}
+		}
+		lanes := task.Lanes()
+		if len(lanes) != 1 || lanes[0] == 0 {
+			// can't happen, really
+			continue
+		}
+		if task.Status() != state.DoneStatus {
+			// ignore non-successful lane
+			laneSnaps[lanes[0]] = ""
+			continue
+		}
+		if _, ok := laneSnaps[lanes[0]]; ok {
+			// ignore lanes we've already seen
+			continue
+		}
+		var snapsup SnapSetup
+		if err := task.Get("snap-setup", &snapsup); err != nil {
+			continue
+		}
+		laneSnaps[lanes[0]] = snapsup.InstanceName()
+	}
+
+	snapNames := make([]string, 0, len(laneSnaps))
+	for _, name := range laneSnaps {
+		if name == "" {
+			continue
+		}
+		snapNames = append(snapNames, name)
+	}
+	return snapNames
+}
+
+// reRefreshSetup holds the necessary details to re-refresh snaps that need it
+type reRefreshSetup struct {
+	UserID int `json:"user-id,omitempty"`
+	*Flags
+}
+
+var reRefreshUpdateMany = updateManyFiltered
+
+func reRefreshFilter(update *snap.Info, snapst *SnapState) bool {
```

Comment:
A comment might be nice: `//reRefreshFilter returns true if the given update needs a re-refresh` or similar, or maybe a slightly more expressive name (but that might be just me) maybe something like: `needsReRefresh()`